src/cdomains/vectorMatrix.ml

```diff
@@ -0,0 +1,857 @@
+open Prelude.Ana
+module Array = Batteries.Array
+module M = Messages
+
+module type RatOps =
+sig
+  type t
+  val add : t -> t -> t
+  val sub : t -> t -> t
+  val mul : t -> t -> t
+  val div : t -> t -> t
+  val neg : t -> t
+  val abs : t -> t
+  val cmp : t -> t -> int
+  val to_string:  t -> string
+  val of_int: int -> t
+  val get_den: t -> IntOps.BigIntOps.t
+  val get_num: t -> IntOps.BigIntOps.t
+end
+
+module ConvenienceOps (A: RatOps) =
+struct
+  let ( *: ) = A.mul
+  let (+:) = A.add
+  let (-:) = A.sub
+  let (/:) = A.div
+  let (=:) x y = A.cmp x y = 0
+  let (<:) x y = A.cmp x y < 0
+  let (>:) x y = A.cmp x y > 0
+  let (<=:) x y = A.cmp x y <= 0
+  let (>=:) x y = A.cmp x y >= 0
+  let of_int x = A.of_int x
+end
+
+module type Vector =
+sig
+  type num
+  type t
+
+  val show: t -> string
+
+  val equal:  t ->  t -> bool
+
+  val keep_vals: t -> int ->  t
+
+  val remove_val: t -> int ->  t
+
+  val set_val: t -> int -> num ->  t
+
+  val set_val_pt_with: t -> int -> num -> t
+
+  val insert_val: int -> num ->  t ->  t
+
+  val apply_with_c: (num -> num -> num) -> num ->  t ->  t
+
+  val apply_with_c_pt_with: (num -> num -> num) -> num -> t -> t
```

Comment:
I'm still not sure what that means in this context. Apron uses `_with` suffix for imperative/in-place modification and that's what our Apron analysis currently does on master as well, but here many (if not most) also have that `pt`.

Reply:
The issue here was that whether these functions modify the value or not depends on the implementation. Thus, for these `pt` variants it may make a modification in place (e.g. if invoked for a array-based module), or may return a new copy (e.g. if invoked for a list-based module).

Thus, a caller may not rely on either behavior being the one that happens and must a) provide defensive copies if they want the original value to be unmodified  (as it is potentially `with`) and b) use the returned value instead of relying on the modification happening in-place (as it may return a new value).

Reply:
I see. It would be possible to provide unambiguous interfaces by lifting immutable lists using refs, but we can skip that for now.

The bigger question is https://github.com/goblint/analyzer/pull/592#discussion_r1027993185, i.e. whether we even want to keep the inferior list versions around like that. If not, then that would also avoid this issue I guess.

Reply:
Since I now removed the list-based implementations, I managed to also get rid of this ambiguous `_pt_with` things by replacing them with `_with` versions. So now these interfaces follow Apron's and our previous conventions.
Moreover, it minimizes the diffs as well (e.g. `ApronAnalysis` on master vs `RelationAnalysis` in this PR) – it's just renames, which makes such large changes easily reviewable.